src/table.js

```diff
@@ -651,6 +649,55 @@ export function __table(source, operations) {
     // operations.types, above.
     source = source.map(d => coerceRow(d, types, schema));
   }
+  return {source, schema};
+}
+
+function applyNames(source, operations) {
+  if (!operations.names) return source;
+  const overridesByName = new Map(operations.names.map((n) => [n.column, n]));
+  return source.map((d) =>
+    Object.fromEntries(Object.keys(d).map((k) => {
+      const override = overridesByName.get(k);
+      return [override?.name ?? k, d[k]];
+    }))
+  );
+}
+
+// This function applies table cell operations to an in-memory table (array of
+// objects); it should be equivalent to the corresponding SQL query. TODO Use
+// DuckDBClient for data arrays, too, and then we wouldn’t need our own __table
+// function to do table operations on in-memory data?
+export function __table(source, operations) {
+  const input = source;
+  let {columns} = source;
+  const typed = applyTypes(source, operations);
+  source = typed.source;
+  let schema = typed.schema;
+  if (operations.derive) {
+    // Derived columns may depend on coerced values from the original data source,
+    // so we must evaluate derivations after the initial inference and coercion
+    // step.
+    const derivedSource = [];
+    operations.derive.map(({name, value}) => {
+      // Derived column formulas may reference renamed columns, so we must
+      // compute derivations on the renamed source.
+      // TODO Allow derived columns to reference other derived columns.
+      applyNames(source, operations).map((row, index, rows) => {
+        const resolved = value(row, index, rows);
+        if (derivedSource[index]) {
+          derivedSource[index] = {...derivedSource[index], [name]: resolved};
+        } else {
+          derivedSource.push({[name]: resolved});
+        }
+      });
```

Comment:
It's possible this is my misunderstanding, but based on this bit here, I expected my derived column values to remain intact after re-naming the column they were based on. Quick demo of what happened instead:

https://github.com/observablehq/stdlib/assets/111310561/9f2693db-ce5f-495f-8eb6-09ebe97aea77

Reply:
oooh that's a great catch!! unfortunately this bit of code doesn't support that case – it makes sure that the column derivations run on the renamed set of columns, but doesn't make sure that the column names in the formula are up-to-date. to make that work, we'd have to modify the formula to replace all instances of the old name with the new name at the time of renaming the column. i can add that to the list of things to fix (it'll require a monorepo change, so shouldn't affect stdlib stuff).

Reply:
Oh, of course! I see that now. Thanks.